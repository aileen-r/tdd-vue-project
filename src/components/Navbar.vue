<template>
  <b-navbar toggleable="lg" type="dark" variant="info">
    <b-navbar-brand id="brand" :to="'/'">VueJS Sandbox</b-navbar-brand>

    <b-navbar-toggle target="nav_collapse" class="mobile-nav-toggle"/>

    <b-collapse is-nav id="nav_collapse">
      <b-navbar-nav class="nav-items">
        <b-nav-item v-for="route in links" :key="route.key" :to="route.url" class="nav-item">
          {{ route.text }}
        </b-nav-item>
      </b-navbar-nav>
    </b-collapse>
  </b-navbar>
</template>

<script>
export default {
  name: 'Navbar',

  data() {
    return {
      links: []
    }
  },

  mounted() {
    this.links = this.$router.options.routes.map(route => {
      return {
        key: route.path.replace('/', ''),
        text: route.name,
        url: route.path 
      };
    });
  }
}
</script>