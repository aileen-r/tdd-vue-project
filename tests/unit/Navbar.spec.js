import { mount, createLocalVue } from '@vue/test-utils';
import BootstrapVue from 'bootstrap-vue';

// import Router from 'vue-router';

import Navbar from '@/components/Navbar';

const localVue = createLocalVue();
localVue.use(BootstrapVue);
// localVue.use(Router);

describe('Navbar', () => {
  let route;
  let routes;
  let router;
  let links;

  // build function suggested over beforeEach() method
  const build = () => {
    const wrapper = mount(Navbar, {
      localVue,
      mocks: {
        $router: router
      }
      // localVue,
      // router
      // propsData: props
    });

    return {
      wrapper,
      brand: () => wrapper.find('#brand'),
      toggle: () => wrapper.find('.mobile-nav-toggle'),
      navItems: () => wrapper.findAll('.nav-items')
    };
  };

  beforeEach(() => {
    route = {
      path: '/page',
      name: 'Page'
    };
    routes = [route];
    router = {
      options: {
        routes
      }
    };
    links = [
      {
        key: route.path.replace('/', ''),
        text: route.name,
        url: route.path
      }
    ]
    // props = { linkUrl: 'https://url.com/hello' };
  });

  // does the component render
  it('renders the component', () => {
    const { wrapper } = build();
    expect(wrapper.html()).toMatchSnapshot();
  });

  // does the component render the right thing
  it('renders branding, collapsible nav and toggle', () => {
    // arrange
    const { brand, toggle, navItems } = build();
    // assert
    expect(brand().exists()).toBe(true);
    expect(toggle().exists()).toBe(true);
    expect(navItems().exists()).toBe(true);
  });

  // is the nav list populated with all router views
  it('populates nav list with links to all router views', () => {
    const { wrapper, navItems } = build();
    expect(wrapper.vm.links).toEqual(links);
    // expect(navItems().
    // TODO - stub router-link to remove error
    // figure out how to find how WrapperArray contains items
  });

  // are active router links highlighted?
  // it('highlights the active route', () => {
  //   const { wrapper, navItems } = build();
  //   console.log(navItems()[0]);
  //   // expect(navItems().contains(route.name)).toBe(true);
  //   expect(wrapper.vm.links).toEqual(links);
  // });

  // does the brand link to homepage?

  // does click even init route change?

  // does nav toggle render only on tablet/mobile
    // how do we test media queries?

  // does nav toggle click toggle nav-items display

});
