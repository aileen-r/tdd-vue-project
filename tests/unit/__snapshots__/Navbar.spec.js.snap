// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders the component 1`] = `
<nav class="navbar navbar-dark bg-info navbar-expand-lg">
  <router-link id="brand" target="_self" class="navbar-brand">VueJS Sandbox</router-link> <button type="button" aria-label="Toggle navigation" aria-controls="nav_collapse" aria-expanded="false" class="mobile-nav-toggle navbar-toggler"><span class="navbar-toggler-icon"></span></button>
  <div id="nav_collapse" class="navbar-collapse collapse" style="display: none;">
    <ul class="navbar-nav nav-items"></ul>
  </div>
</nav>
`;
